File: test_src/cpp/hierarchical_parallelism/atomic/double/target__teams__distribute__parallel__loop.cpp
#include <iostream>
#include <cmath>
#include <stdexcept>
#
bool almost_equal(double x, double gold, float tol) {
        return gold * (1-tol) <= x && x <= gold * ( 1+tol );
}
void test_target__teams__distribute__parallel__loop(){
 // Input and Outputs
 const int L = 4096;
 const int M = 64;
double counter{};
// Main program
#pragma omp target  map(tofrom:counter) 
{
#pragma omp teams 
{
#pragma omp distribute 
    for (int i = 0 ; i < L ; i++ )
{
#pragma omp parallel 
{
#pragma omp loop 
    for (int j = 0 ; j < M ; j++ )
{
#pragma omp atomic update
counter += double { 1 };
    } 
    } 
    } 
    } 
    } 
// Validation
if ( !almost_equal(counter,double { L*M }, 0.1)  ) {
    std::cerr << "Expected: " << L*M << " Got: " << counter << std::endl;
    throw std::runtime_error( "target__teams__distribute__parallel__loop give incorect value when offloaded");
}
}
int main()
{
    test_target__teams__distribute__parallel__loop();
}

File: test_src/fortran/hierarchical_parallelism/atomic/double_precision/target_teams__distribute__parallel__loop.f90


program target_teams__distribute__parallel__loop


    implicit none
  
    INTEGER :: L = 5
    INTEGER :: i
    INTEGER :: M = 6
    INTEGER :: j
    
    DOUBLE PRECISION :: COUNTER = 0

    
    

     
    
    !$OMP TARGET TEAMS   MAP(TOFROM: COUNTER) 



    

    
    !$OMP DISTRIBUTE 


    DO i = 1 , L 


    

    
    !$OMP PARALLEL 



    

    
    !$OMP LOOP 


    DO j = 1 , M 


    

    

!$OMP ATOMIC UPDATE

counter = counter + 1.


 
     

    END DO

    !$OMP END LOOP
     

    !$OMP END PARALLEL
     

    END DO

    !$OMP END DISTRIBUTE
     

    !$OMP END TARGET TEAMS
    

    IF  ( ( ABS(COUNTER - L*M) ) > 10*EPSILON(COUNTER) ) THEN
        write(*,*)  'Expected L*M Got', COUNTER
        call exit(1)
    ENDIF

end program target_teams__distribute__parallel__loop